MdePkg/Include/Library/UefiUsbLib.h

```diff
@@ -552,4 +553,123 @@ UsbClearEndpointHalt (
   OUT UINT32               *Status
   );
 
+/**
+  Retrieve the interface descriptor details from the interface setting.
+
+  This is an extended version of UsbIo->GetInterfaceDescriptor. It returns the interface
+  descriptor for an alternate setting of the interface without executing SET_INTERFACE
+  transfer. It also returns the number of class specific interfaces.
+
+  @param[in]  This              A pointer to the EFI_USB_IO_PROTOCOL instance.
+  @param[in]  Setting           Interface alternate setting.
+  @param[out]  Descriptor       The caller allocated buffer to return the contents of the Interface descriptor.
+  @param[out]  CsInterfaceNumber  Number of class specific interfaces for this interface setting.
+
+  @retval EFI_SUCCESS           Output parameters were updated successfully.
+  @retval EFI_INVALID_PARAMETER Descriptor or CsInterfaceNumber is NULL.
+  @retval EFI_UNSUPPORTED       Setting is greater than the number of alternate settings in this interface.
+  @retval EFI_DEVICE_ERROR      Error reading device data.
+
+**/
+EFI_STATUS
+EFIAPI
+UsbGetInterfaceDescriptorSetting (
+  IN  EFI_USB_IO_PROTOCOL           *This,
+  IN  UINTN                         Setting,

```

Comment:
This value has to be related to the USB I/O Protocol or USB Specification.  Is there a better type than UINTN so the caller of these lib services know what Setting values are valid?

Reply:
Good point. USB_INTERFACE_DESCRIPTOR.AlternateSetting is UINT8. Done.

Reply:
Just changing to UINT8 does not help.  What are the legal values for Setting?  What is the behavior of the API for each value of Setting?  Is there an enum or #defines for the values?

Reply:
USB specification does not define legal values for the Setting. It is merely an index of the alternate interface reported by the device. Here is how it is described in USB 2.0 Spec, chapter 9.2.3:
*As part of the configuration process, the host sets the device configuration and, where necessary, selects the
appropriate alternate settings for the interfaces.*
The setting definitions, enum will be on the device driver side, not on this library's API.

For the functions that are already part of this library AlternateSetting argument is UINT16. I think this is wrong, because Alternate setting is a part of the interface descriptor and defined in Usb.h as follows:
`///
/// Standard Interface Descriptor
/// USB 2.0 spec, Section 9.6.5
///
typedef struct {
  UINT8    Length;
  UINT8    DescriptorType;
  UINT8    InterfaceNumber;
  UINT8    AlternateSetting;
  UINT8    NumEndpoints;
  UINT8    InterfaceClass;
  UINT8    InterfaceSubClass;
  UINT8    InterfaceProtocol;
  UINT8    Interface;
} USB_INTERFACE_DESCRIPTOR;
`
I thought of a different PR for this fix but can be addressed here as well, please advise.